src/rust/grapl-service/src/decoder/ndjson.rs

```diff
@@ -1,31 +1,67 @@
-use std::sync::Arc;
-
 use serde::Deserialize;
-use sqs_executor::event_decoder::PayloadDecoder;
+use sqs_executor::{
+    errors::{
+        CheckedError,
+        Recoverable,
+    },
+    event_decoder::PayloadDecoder,
+};
 
 use crate::decoder::decompress::PayloadDecompressionError;
 
 #[derive(Debug, Clone, Default)]
 pub struct NdjsonDecoder;
 
-impl<D> PayloadDecoder<Vec<Result<D, Arc<serde_json::Error>>>> for NdjsonDecoder
+#[derive(thiserror::Error, Debug)]
+pub enum NdjsonDecoderError {
+    #[error("DecompressionError")]
+    Decompression(#[from] PayloadDecompressionError),
+    #[error("DeserializeError")]
+    Deserialization(#[from] serde_json::Error),
+}
+
+impl CheckedError for NdjsonDecoderError {
+    fn error_type(&self) -> Recoverable {
+        match self {
+            Self::Decompression(_) => Recoverable::Persistent,
+            Self::Deserialization(_) => Recoverable::Persistent,
+        }
+    }
+}
+
+impl<D> PayloadDecoder<Vec<D>> for NdjsonDecoder
 where
     for<'a> D: Deserialize<'a>,
 {
-    type DecoderError = PayloadDecompressionError;
+    type DecoderError = NdjsonDecoderError;
 
-    fn decode(
-        &mut self,
-        body: Vec<u8>,
-    ) -> Result<Vec<Result<D, Arc<serde_json::Error>>>, Self::DecoderError> {
+    fn decode(&mut self, body: Vec<u8>) -> Result<Vec<D>, Self::DecoderError> {
         let decompressed = super::decompress::maybe_decompress(body.as_slice())?;
 
-        let results: Vec<_> = decompressed
+        let mut first_deserialization_error: Option<serde_json::Error> = None;
+
+        let events: Vec<_> = decompressed
             .split(|byte| *byte == b'\n')
             .filter(|line| !line.is_empty())
-            .map(|line| serde_json::from_slice(line).map_err(|err| Arc::new(err)))
+            .filter_map(|line| {
+                let result = serde_json::from_slice(line);
+                match result {
+                    Ok(result) => Some(result),
+                    Err(error) => {
+                        tracing::error!(message="Unable to deserialize OSQuery event.", error=?error);
+
+                        if first_deserialization_error.is_none() {
+                            first_deserialization_error = Some(error);
+                        }
+                        None
+                    }
+                }
+            })
             .collect();
 
-        Ok(results)
+        match first_deserialization_error {
+            Some(error) if events.is_empty() => Err(error.into()),
```

Comment:
Can we just add a comment here? The interfaces in sqs-executor are so bad it seems worth pointing out that the in the case where events is *not* empty we still want to make forward progress, so we just log the errors but otherwise pass what we can forward.